src/Planner/Planner.cpp

```diff
@@ -720,7 +720,7 @@ bool addPreliminaryLimitOptimizationStepIfNeeded(QueryPlan & query_plan,
 
     bool apply_limit = query_processing_info.getToStage() != QueryProcessingStage::WithMergeableStateAfterAggregation;
     bool apply_prelimit = apply_limit &&
-        query_node.hasLimit() &&
+        query_analysis_result.limit_length &&
```

Comment:
Unrelated changes in this file.

Reply:
it's a bug fix - limit should be taken here from analysis result

Reply:
Where was bug ?
```
if (query_node.hasLimit())
{
    /// Constness of limit is validated during query analysis stage
    limit_length = query_node.getLimit()->as<ConstantNode &>().getValue().safeGet<UInt64>();
}
```
We should not use this if we want to check if query node has limit.

Reply:
the use of `query_node.hasLimit()` is a bug - further we a calling a`ddPreliminaryLimitStep` passing `query_analysis_result` and using `query_analysis_result.limit_length`:
```
void addPreliminaryLimitStep(QueryPlan & query_plan,
    const QueryAnalysisResult & query_analysis_result,
    const PlannerContextPtr & planner_context,
    bool do_not_skip_offset)
{
    UInt64 limit_offset = query_analysis_result.limit_offset;
    UInt64 limit_length = query_analysis_result.limit_length;
...
```
so, apparently we should use query_analysis_result.limit_length there